src/boot/efi/secure-boot.c

```diff
@@ -35,3 +37,87 @@ static const char sbat[] _used_ _section_(".sbat") =
         SBAT_PROJECT ",1,The systemd Developers," SBAT_PROJECT "," PROJECT_VERSION "," PROJECT_URL "\n"
         SBAT_PROJECT "." SBAT_DISTRO "," STRINGIFY(SBAT_DISTRO_GENERATION) "," SBAT_DISTRO_SUMMARY "," SBAT_DISTRO_PKGNAME "," SBAT_DISTRO_VERSION "," SBAT_DISTRO_URL "\n";
 #endif
+
+EFI_STATUS secure_boot_enroll(EFI_FILE *root_dir, Config *config, ConfigEntry *entry) {
+        assert(root_dir);
+        assert(config);
+        assert(entry);
+
+        EFI_STATUS err = EFI_SUCCESS;
+
+        clear_screen(COLOR_NORMAL);
+
+        Print(L"Enrolling secure boot keys from directory: \\loader\\keys\\%s\n"
+              L"Warning: Enrolling custom Secure Boot keys might soft-brick your machine!\n",
+              entry->path);
+
+        UINT32 timeout_sec = 15;
+        for(;;) {
+                PrintAt(0, ST->ConOut->Mode->CursorRow, L"Enrolling in %2u s, press any key to abort.", timeout_sec);
+
+                UINT64 key;
+                err = console_key_read(&key, 1000 * 1000);
+                if (err == EFI_NOT_READY)
+                        continue;
+                if (err == EFI_TIMEOUT) {
+                        if (timeout_sec == 0) /* continue enrolling keys */
+                                break;
+                        timeout_sec--;
+                        continue;
+                }
+                if (EFI_ERROR(err))
+                        return log_error_status_stall(err, L"Error waiting for user input to enroll Secure Boot keys: %r", err);
+
+                /* user aborted */
+                return EFI_SUCCESS;
+        }
+
+        _cleanup_(file_closep) EFI_FILE *dir = NULL;
+        UINT32 sb_vars_opts =
+                EFI_VARIABLE_NON_VOLATILE |
+                EFI_VARIABLE_BOOTSERVICE_ACCESS |
+                EFI_VARIABLE_RUNTIME_ACCESS |
+                EFI_VARIABLE_TIME_BASED_AUTHENTICATED_WRITE_ACCESS;
+
+        err = open_directory(root_dir, entry->path, &dir);
+        if (EFI_ERROR(err))
+                return log_error_status_stall(err, L"Failed opening keys directory %s: %r", entry->path, err);
+
+        struct {
+                const CHAR16 *name;
+                const CHAR16 *filename;
+                const EFI_GUID vendor;
+                CHAR8 *buffer;
+                UINTN size;
+        } sb_vars[] = {
+                { L"db",  L"db.esl",  EFI_IMAGE_SECURITY_DATABASE_VARIABLE, NULL, 0 },
+                { L"KEK", L"KEK.esl", EFI_GLOBAL_VARIABLE, NULL, 0 },
+                { L"PK",  L"PK.esl",  EFI_GLOBAL_VARIABLE, NULL, 0 },
+        };
+
+        for (UINTN i = 0; i < ELEMENTSOF(sb_vars); i++) {
+                err = file_read(dir, sb_vars[i].filename, 0, 0, &sb_vars[i].buffer, &sb_vars[i].size);
+                if (EFI_ERROR(err)) {
+                        log_error_stall(L"Failed reading file %s: %r", sb_vars[i].filename, err);
```

Comment:
This should give the full path:
```suggestion
                        log_error_stall(L"Failed reading file %s\\%s: %r", entry->path, sb_vars[i].filename, err);
```

Reply:
Not addressed.